google/cloud/spanner/internal/logging_result_set_reader_test.cc

```diff
@@ -0,0 +1,114 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/logging_result_set_reader.h"
+#include "google/cloud/spanner/testing/mock_partial_result_set_reader.h"
+#include "google/cloud/internal/make_unique.h"
+#include "google/cloud/log.h"
+#include "google/cloud/testing_util/assert_ok.h"
+#include "google/cloud/testing_util/capture_log_lines_backend.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+namespace {
+
+namespace spanner_proto = google::spanner::v1;
+
+class LoggingResultSetReaderTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    backend_ =
+        std::make_shared<google::cloud::testing_util::CaptureLogLinesBackend>();
+    logger_id_ = google::cloud::LogSink::Instance().AddBackend(backend_);
+  }
+
+  void TearDown() override {
+    google::cloud::LogSink::Instance().RemoveBackend(logger_id_);
+    logger_id_ = 0;
+  }
+
+  void HasLogLineWith(std::string const& contents) {
+    auto count =
+        std::count_if(backend_->log_lines.begin(), backend_->log_lines.end(),
+                      [&contents](std::string const& line) {
+                        return std::string::npos != line.find(contents);
+                      });
+    EXPECT_NE(0, count);
+  }
+
+ private:
+  std::shared_ptr<google::cloud::testing_util::CaptureLogLinesBackend> backend_;
+  long logger_id_ = 0;  // NOLINT
+};
+
+TEST_F(LoggingResultSetReaderTest, TryCancel) {
+  auto mock = google::cloud::internal::make_unique<
+      spanner_testing::MockPartialResultSetReader>();
+  EXPECT_CALL(*mock, TryCancel()).Times(1);
+  LoggingResultSetReader reader(std::move(mock));
+  reader.TryCancel();
+
+  HasLogLineWith("TryCancel");
+}
+
+TEST_F(LoggingResultSetReaderTest, Read) {
+  auto mock = google::cloud::internal::make_unique<
+      spanner_testing::MockPartialResultSetReader>();
+  EXPECT_CALL(*mock, Read())
+      .WillOnce([] {
+        spanner_proto::PartialResultSet result;
+        result.set_resume_token("test-token");
+        return result;
+      })
+      .WillOnce([] {
+        return google::cloud::optional<spanner_proto::PartialResultSet>{};
+      });
+  LoggingResultSetReader reader(std::move(mock));
+  auto result = reader.Read();
+  ASSERT_TRUE(result.has_value());
+  EXPECT_EQ("test-token", result->resume_token());
+
+  HasLogLineWith("Read");
+  HasLogLineWith("test-token");
+
+  result = reader.Read();
+  ASSERT_FALSE(result.has_value());
+  HasLogLineWith("(optional-with-no-value)");
```

Comment:
Something to consider for the future: Should we have some way to make the expectation only apply to log lines accumulated since the last expectation?

Reply:
Done.